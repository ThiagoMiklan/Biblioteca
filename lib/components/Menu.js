"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var React = _interopRequireWildcard(require("react"));

var _type_validations = require("../tools/type_validations.js");

var props_obj = {
  definition: "string"
};

var Menu = function Menu(props) {
  (0, _type_validations.validate)(props, props_obj, "Menu");
  var definition = props.definition == undefined ? "menu" : "menu " + props.definition;
  return /*#__PURE__*/React.createElement("aside", {
    className: definition
  }, props.children);
};

var _default = Menu;
exports["default"] = _default;