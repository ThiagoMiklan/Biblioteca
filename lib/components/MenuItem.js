"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _type_validations = require("../tools/type_validations");

var props_obj = {
  itens: "array",
  label: "string"
};

var MenuItem = function MenuItem(props) {
  (0, _type_validations.validate)(props, props_obj, "MenuItem");
  return /*#__PURE__*/_react["default"].createElement("div", null, props.itens != undefined && /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, assembleLabel(props.label), /*#__PURE__*/_react["default"].createElement("ul", {
    className: "menu-list"
  }, props.itens != undefined && assembleItem(props.itens))));
};

function assembleLabel(label) {
  return /*#__PURE__*/_react["default"].createElement("p", {
    "class": "menu-label"
  }, label);
}
/*

*/


function assembleItem(itens) {
  var item_code = "";

  if (Array.isArray(itens)) {
    item_code = assembleItemWithSubList(itens);
  } else {
    item_code = assembleSingleItem(itens);
  }

  return item_code;
}

function assembleSingleItem(item) {
  var item_code = "";

  if (item["sub_itens"] != undefined) {
    item_code = decideTypeOfItem(item);
  } else {
    var definition = item["definition"];
    item_code = /*#__PURE__*/_react["default"].createElement("li", null, /*#__PURE__*/_react["default"].createElement("a", {
      className: definition,
      onClick: item["onClick"]
    }, item["value"]));
  }

  return item_code;
}

function assembleSubItens(item) {
  var definition = item["definition"];
  return /*#__PURE__*/_react["default"].createElement("li", null, /*#__PURE__*/_react["default"].createElement("a", {
    className: definition,
    onClick: item["onClick"]
  }, item["value"]));
}

function assembleItemWithSubList(itens) {
  return itens.map(function (item) {
    return decideTypeOfItem(item);
  });
}

function decideTypeOfItem(item) {
  var item_code = "";

  if (item["sub_itens"] != undefined) {
    var definition = item["definition"];
    item_code = /*#__PURE__*/_react["default"].createElement("li", null, /*#__PURE__*/_react["default"].createElement("a", {
      className: definition,
      onClick: item["onClick"]
    }, " ", item["value"]), /*#__PURE__*/_react["default"].createElement("ul", null, item["sub_itens"].map(function (item) {
      return assembleSubItens(item);
    })));
  } else {
    item_code = assembleSingleItem(item);
  }

  return item_code;
}

var _default = MenuItem;
exports["default"] = _default;