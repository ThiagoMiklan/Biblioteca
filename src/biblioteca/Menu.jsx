// @flow
import * as React from 'react';
import PropTypes from 'prop-types';

type Props = {
    children?: React.Node
}

const Menu = (props:Props) =>{
    var definition = (props.definition == undefined) ? "menu" : "menu "+ props.definition;
        return(
            <aside className={definition}>
              {props.children}
            </aside>
        );
}


Menu.propTypes = {
    menu_itens: PropTypes.array,
    sub_itens: PropTypes.array
}

export default Menu;