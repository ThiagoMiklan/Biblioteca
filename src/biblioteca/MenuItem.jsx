// @flow
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import getClassName from '../tools/getClassName';
import checkerDefinition from '../tools/checker.js';

/*
    Menu Item foi desenvolvido para representar uma parte de um menu, 
    representando não somente o menu em si, mas também o label de cada item/itens

    Funcionamento: Para criar um menu é possível passar como props.itens um objeto ou um array
    Objeto -> Informar quando for necessário somente um item
    Ex: 
     GENERAL -> label
        Manage your team -> item
            Member  -> subitem
            Plugin  -> subitem
            Add a member -> subitem

    Nota-se que nesse caso é informado junto com o item, seus subitens(que são opcionais)

    Array-> Informar quando necessário mais de um item
    Ex:
    ADMINISTRATION -> label
        Dashboard -> item
        Customers -> item

    Existem a possibilidade de mesclar itens com subitens



*/
type Props = {
    itens?: Array<Object>,
    label?: string
}

const MenuItem = (props:Props)=>{
    
    return (
        
            <div>
                {props.itens != undefined && 
                 <>
                 {assembleLabel(props.label)}
                 <ul className="menu-list">
                 {props.itens != undefined && assembleItem(props.itens)}
                 </ul>
                 </>
                }
            </div>
            );
}

function assembleLabel(label){
    return <p class="menu-label">
            {label}
           </p>;
}
/*

*/
function assembleItem(itens) {
    var item_code = "";
    // item é um array²z
    if (Array.isArray(itens)) {
        // [sim] = monta o item baseado em um array
       item_code = assembleItemWithSubList(itens);
    } else {
        // [não] = monta um item normal, o item mais simples
       item_code = assembleSingleItem(itens);
    }

    return item_code;
}

// caso simples onde apenas um item é desejado, informando apenas um objeto
function assembleSingleItem(item) {
    var item_code = "";
  
    if(item["sub_itens"]!= undefined){
        item_code = decideTypeOfItem(item);
    }else{
        checkerDefinition(item["definition"],"Menu");
        var definition =  classnames(getClassName(item["definition"],"Menu"));
        item_code = <li><a className={definition} onClick={item["onClick"]} >{item["value"]}</a></li>; 
    }
    return item_code;
}
function assembleSubItens(item){
    checkerDefinition(item["definition"],"Menu");
    var definition = classnames(getClassName(item["definition"],"Menu"));
    return <li><a className={definition} onClick={item["onClick"]}>{item["value"]}</a></li>;
}

// caso mais complexo onde um item possui uma sublista
function assembleItemWithSubList(itens) {
    return (itens.map(item => decideTypeOfItem(item)));
}

// método usado para retornar o código referente a um item de um array de forma simples
// ou um item de um array que contenha uma sublista
function decideTypeOfItem(item) {
    var item_code = "";
    
    // se possuir sub_itens
    if (item["sub_itens"] != undefined) {
        checkerDefinition(item["definition"],"Menu");
        var definition =  classnames(getClassName(item["definition"],"Menu"));
        item_code = <li>
            <a className={definition} onClick={item["onClick"]}> {item["value"]}</a>
            <ul>
                {item["sub_itens"].map(item => assembleSubItens(item))}
            </ul>
        </li>
    } else {
        item_code = assembleSingleItem(item);
    }

    return item_code;
}

MenuItem.propTypes ={
    label: PropTypes.string,
    itens: PropTypes.array
}

export default MenuItem;