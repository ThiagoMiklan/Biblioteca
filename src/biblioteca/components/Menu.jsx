import React from 'react';
import PropTypes from 'prop-types';
import classname from 'classnames';
import getClassName from '../../tools/getClassName';
import Button from '../elements/Button.jsx';
import { renderToString } from 'react-dom/server';

class Menu extends React.Component{
    
    render(){
        return(
            <aside class="menu">
                
            </aside>

        );
    }

}


Menu.propTypes = {
    menu_itens: PropTypes.array,
    sub_itens: PropTypes.array
}

export default Menu;