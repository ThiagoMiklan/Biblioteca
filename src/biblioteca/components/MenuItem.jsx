import React from 'react';
import PropTypes from 'prop-types';
import classname from 'classnames';
import getClassName from '../../tools/getClassName';
import Button from '../elements/Button.jsx';
import { renderToString } from 'react-dom/server';

class MenuItem extends React.Component{

    render(){
        return(
          <div>
            {assembleItem(this.props.item, this.props.sub_itens)}
          </div>);
    }
}


function assembleItem(item,sub_itens){
    if(sub_itens != undefined && Array.isArray(sub_itens)){
        assembleItemWithSubList(item,sub_itens);
    }else{

    }
}
function assembleItemWhitoutSubItem(item){
   return <li><a>{item}</a></li>;
}

function assembleItemWithSubList(item,sub_itens){
    
    return (<li>
        <a> {item}</a>
        <ul>
            {sub_itens.map(item=> assembleItemWhitoutSubItem(item))}
        </ul>
    </li>);

    
}


export default MenuItem;