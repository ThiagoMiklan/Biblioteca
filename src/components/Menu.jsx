// @flow
import * as React from 'react';
import { validate } from '../tools/type_validations.js';

type Props = {
    children?: React.Node,
    definition?: string
}

const props_obj = {
    definition: "string"
}

const Menu = (props:Props) =>{
    validate(props, props_obj,"Menu");
    var definition = (props.definition == undefined) ? "menu" : "menu "+ props.definition;
        return(
            <aside className={definition}>
              {props.children}
            </aside>
        );
}

export default Menu;  