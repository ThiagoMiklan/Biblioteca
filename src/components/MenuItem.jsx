// @flow
import React from 'react';
import { validate } from '../tools/type_validations';
import {MenuItemItem} from '../tools/types';

type Props = {
    itens?: Array<MenuItemItem>,
    label?: string
}

const props_obj = {
    itens: "array",
    label: "string"
}

const MenuItem = (props:Props)=>{
    validate(props,props_obj,"MenuItem");
    return (
        
            <div>
                {props.itens != undefined && 
                 <>
                 {assembleLabel(props.label)}
                 <ul className="menu-list">
                 {props.itens != undefined && assembleItem(props.itens)}
                 </ul>
                 </>
                }
            </div>
            );
}

function assembleLabel(label){
    return <p class="menu-label">
            {label}
           </p>;
}
/*

*/
function assembleItem(itens) {
    var item_code = "";
   
    if (Array.isArray(itens)) {
       item_code = assembleItemWithSubList(itens);
    } else {
       item_code = assembleSingleItem(itens);
    }

    return item_code;
}


function assembleSingleItem(item) {
    var item_code = "";
  
    if(item["sub_itens"]!= undefined){
        item_code = decideTypeOfItem(item);
    }else{
     
        var definition =  item["definition"];
        item_code = <li><a className={definition} onClick={item["onClick"]} >{item["value"]}</a></li>; 
    }
    return item_code;
}
function assembleSubItens(item){
 
    var definition= item["definition"];
    return <li><a className={definition} onClick={item["onClick"]}>{item["value"]}</a></li>;
}


function assembleItemWithSubList(itens) {
    return (itens.map(item => decideTypeOfItem(item)));
}

function decideTypeOfItem(item) {
    var item_code = "";
    
    if (item["sub_itens"] != undefined) {
      
        var definition = item["definition"];
        item_code = <li>
            <a className={definition} onClick={item["onClick"]}> {item["value"]}</a>
            <ul>
                {item["sub_itens"].map(item => assembleSubItens(item))}
            </ul>
        </li>
    } else {
        item_code = assembleSingleItem(item);
    }

    return item_code;
}


export default MenuItem;