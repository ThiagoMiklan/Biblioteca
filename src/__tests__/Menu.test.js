import React from 'react';
import Menu from '../biblioteca/Menu.jsx';
import MenuItem from '../biblioteca/MenuItem.jsx';
import renderer from 'react-test-renderer';
 

var onClickItem = ()=>{alert("Clicou")};

var itens_general = [
    {value: "General"}, {value:"Dashboard"}
]

var itens_adm = [
    {value:"Team Settings"},{value:"Manage your team",definition:"active",
    sub_itens:[{value:"Members"},{value:"Plugins"},{value:"Add a member"}]},
    {value:"Invitations"},{value:"Cloud Storage Environment Settings"},
    {value:"Authentication"}
]

var itens_tran =[
    {value: "Payments"},{value: "Transfer"},{value:"Balance"}
]

var item = {};

// para adicionar os subitens, informa o nome do iteam, o label e os itens de subitem
var itens_general_without_event = {value:"Manage your team", 
sub_itens:[{value:"Member"},{value:"Plugin"},{value:"Add a member"}]
}

var itens_administration_without_event =
 [{value:"Dashboard"}, 
 {value:"Customers",
 sub_itens:[{value:"Member"},
 {value:"Plugin"},
 {value:"Add a member"}]}];
 
var itens_merge = [{value:"item 1",sub_itens:[{value:"1"},{value:"2"}]} ,{value:"item 2",sub_itens:[{value:1},{value:2}]}]


var itens_general_with_event = 
{value:"Manage your team",onClick:onClickItem ,sub_itens:[{value:"Member", onClick:onClickItem},{value:"Plugin"},{value:"Add a member"}]}

var itens_administration_with_event =
[{value:"Dashboard",onClick:onClickItem },
 {value:"Customers",sub_itens:["Member","Plugin","Add a member"]},
];


test('menu',()=>{
    const menu =  renderer.create(
        <Menu>
                <MenuItem itens={itens_general} label="General"></MenuItem>
                <MenuItem itens={itens_adm} label="Administration"></MenuItem>
                <MenuItem itens={itens_tran} label="Transactions"></MenuItem>
          </Menu>
    ).toJSON();

    expect(menu).toMatchSnapshot();
})



