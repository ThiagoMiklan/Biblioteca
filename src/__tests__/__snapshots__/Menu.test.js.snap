// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menu 1`] = `
<aside
  class="menu"
>
  <div>
    <p
      class="menu-label"
    >
      General
    </p>
    <ul
      className="menu-list"
    >
      <li>
        <a
          className=""
        >
          General
        </a>
      </li>
      <li>
        <a
          className=""
        >
          Dashboard
        </a>
      </li>
    </ul>
  </div>
  <div>
    <p
      class="menu-label"
    >
      Administration
    </p>
    <ul
      className="menu-list"
    >
      <li>
        <a
          className=""
        >
          Team Settings
        </a>
      </li>
      <li>
        <a
          className="is-active"
        >
           
          Manage your team
        </a>
        <ul>
          <li>
            <a
              className=""
            >
              Members
            </a>
          </li>
          <li>
            <a
              className=""
            >
              Plugins
            </a>
          </li>
          <li>
            <a
              className=""
            >
              Add a member
            </a>
          </li>
        </ul>
      </li>
      <li>
        <a
          className=""
        >
          Invitations
        </a>
      </li>
      <li>
        <a
          className=""
        >
          Cloud Storage Environment Settings
        </a>
      </li>
      <li>
        <a
          className=""
        >
          Authentication
        </a>
      </li>
    </ul>
  </div>
  <div>
    <p
      class="menu-label"
    >
      Transactions
    </p>
    <ul
      className="menu-list"
    >
      <li>
        <a
          className=""
        >
          Payments
        </a>
      </li>
      <li>
        <a
          className=""
        >
          Transfer
        </a>
      </li>
      <li>
        <a
          className=""
        >
          Balance
        </a>
      </li>
    </ul>
  </div>
</aside>
`;
